### DISCUSSION TITLE
wire:stream doesn't work with Octane

### DISCUSSION THREAD

**Discussion by gazzoy** [livewire/livewire#7798] [7. Bugs (archived)]
### Livewire version

v3.4.1

### Laravel version

v10.42.0

### Which PHP version are you using?

PHP 8.2

### Steps To Reproduce

It seems `wire:stream` doesn't work with `Octane`.

In order to reproduce the issue, I created a repo with a simple `CountDown` component described on [Livewire docs](https://livewire.laravel.com/docs/wire-stream) on top of a fresh Laravel(with Sail), Livewire, and Octane(swoole).

1. clone [gazzoy/livewire-stream-octane-issue](https://github.com/gazzoy/livewire-stream-octane-issue)
2. run following commands:
```
$ cd /path/to/livewire-stream-octane-issue
$ docker run --rm \
    -u "$(id -u):$(id -g)" \
    -v "$(pwd):/var/www/html" \
    -w /var/www/html \
    laravelsail/php82-composer:latest \
    composer install --ignore-platform-reqs
$ ./vendor/bin/sail build --no-cache
$ ./vendor/bin/sail up -d
```
3. open `http://localhost/count-down` on brwoser, then click `Start count-down` button.
4. `CountDown` component supposed to be counting down as 3 -> 2 -> 1 -> 0, but with Octane it only counts like 3 -> -1 which I assume Octane cannot detect the Livewire streaming process or something like that.

I suspect that `Livewire\Features\SupportStreaming\HandlesStreaming` might contains a bug, but couldn't figure it out where I should fix.

### Repository or Wirebox URL

https://github.com/gazzoy/livewire-stream-octane-issue

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by gazzoy** [comment]
@lrljoe @DanielSpravtsev @joshhanley @calebporzio Do you guys happen to have have any idea to what is causing this? That would be great if you chould share some hints or thoughts on this.

**Reply by DanielSpravtsev** [reply]
Hey!
I think it could be cached somehow between request.
For this reason you see data changes only on requestStart and requestEnd stream events.
I haven't used Octane with Livewire yet, so I can't be 100% sure of a problem causing this, but you should dig into flushing or persisting between requests. 
Hope I helped a little bit, maybe @joshhanley has more detailed explanation

**Reply by gazzoy** [reply]
> I think it could be cached somehow between request.
> For this reason you see data changes only on requestStart and requestEnd stream events.
> I haven't used Octane with Livewire yet, so I can't be 100% sure of a problem causing this, but you should dig into flushing or persisting between requests.

Thanks! Will look into it 👀

**Reply by joshhanley** [reply]
@gazzoy and @DanielSpravtsev yeah not sure. We haven't yet got around to testing V3 with Octane, so it's likely that something is still being stored in between requests when it shouldn't.

What exactly was the issue - like does the network request fire as expected? Does it finish processing? Does it hang? If you can provide some more details on what is/isn't happening, we might be able to narrow it down.

**Reply by gazzoy** [reply]
> What exactly was the issue - like does the network request fire as expected? Does it finish processing? Does it hang? If you can provide some more details on what is/isn't happening, we might be able to narrow it down.

@joshhanley Thank you for your reply. Got it! Will do some debugging to find out.

**Reply by gazzoy** [reply]
@joshhanley 

Here's what I've found. To recap, the issue seems to be how we return a `StreamedResponse` rather than Octane caching or flushing problem. But to be honest, I'm a bit confused as the issue seems to be due to `Octane` bug, but at the same time, we possibily can say its Livewire's bug. Let me describe some facts I found in two cases: w/o Octane, and w/ Octane.

## Without Octane

In this case, `wire:stream` works perfectly fine.

1. After we call `HandlesStreaming#stream`, in order to initiate the streaming response `Features/SupportStreaming/SupportStreaming.php` generates `StreamedResponse` and sends the headers back.

https://github.com/livewire/livewire/blob/main/src/Features/SupportStreaming/SupportStreaming.php#L22-L29
```
        static::$response = response()->stream(null, 200, [
            'Cache-Control' => 'no-cache',
            'Content-Type' => 'text/event-stream',
            'X-Accel-Buffering' => 'no',
            'X-Livewire-Stream' => true,
        ]);
        static::$response->sendHeaders();
```        

2. As soon as the header sent back, `livewire.js` processes the streaming request in case the response's headers contain `X-Livewire-Stream`. By these processes, I believe we can update the value in blade through the streaming.

https://github.com/livewire/livewire/blob/main/dist/livewire.js#L9273-L9293
```
  on("request", ({ respond }) => {
    respond((mutableObject) => {
      let response = mutableObject.response;
      if (!response.headers.has("X-Livewire-Stream"))
        return;
      mutableObject.response = {
      ...
```


## With Octane

In this case, `wire:stream` doesn't work.

For debuging purpose, I put `Log::debug` in `CountDown` component like this:

```
class CountDown extends Component
{
    public $start = 3;
 
    public function begin()
    {
        while ($this->start >= 0) {
            Log::debug('start:' . $this->start);

            // Stream the current count to the browser...
            $this->stream(  
                to: 'count',
                content: $this->start,
                replace: true,
            );
        ...            
```

1. When we call `HandlesStreaming#stream`, the log shows as below which totally makes sense:

```
start:3
start:2
start:1
start:0
```

2. However, on browser we only see the count down like 3 -> -1 which doesn't make sense.

3. As of I understood, in order to update requests in Octane, we need to complete each request and send the response back. Otherwise Octane wait for the response and never process the next request.
https://github.com/laravel/octane/blob/2.x/src/ApplicationGateway.php#L36

4. Then, the problem is, in case we use Octane, it seeems `StreamedResponse` never send its response to `livewire.js` until all of the streamings would be done. Therefore, `livewire.js` never receive the request which contains `X-Livewire-Stream` in header, hence the count-down on a blade works like 3 -> -1. I assume this could be happened because `wire:stream` only calls `static::$response->sendHeaders();` which Octane cannot recognize as a response for the request, but not sure how we can fix.


I'm at a bit of a dead end... Am I missing something?

**Reply by joshhanley** [reply]
@gazzoy thanks for digging! Hmm yeah I'm not really sure either. I'm not overly familiar with how Octane works under the hood. I don't really understand why it would handle a stream response any different than any other response 🤔

**Reply by gazzoy** [reply]
@joshhanley thanks for your reply! Ok, so at this point, I wouldn't use `wire:stream` with `Octane` then 🤒 Maybe I would look into it again later.

**Reply by joshhanley** [reply]
@gazzoy yep, we've added a warning to the docs for now until it can be worked out PR #7929

**Reply by gazzoy** [reply]
> yep, we've added a warning to the docs for now until it can be worked out PR https://github.com/livewire/livewire/pull/7929

@joshhanley Cool! Thanks for updating the docs.

**Reply by danharrin** [reply]
I have opened an issue on Octane to see if it can be fixed: https://github.com/laravel/octane/issues/903

**Reply by gazzoy** [reply]
@danharrin Thanks! Let's see how it goes 👀

**Reply by Lumethys** [reply]
So it looks like the Octane side has been fixed, any information on this side guys?

**Reply by gazzoy** [reply]
@Lumethys I believe the case running Octane w/ roadrunner has been fixed by https://github.com/laravel/octane/pull/939. However, the other cases, such as using swoole or frankenphp are most likely still having problems.
